### DISCUSSION TITLE
Formatting of visual selection with LSP

### DISCUSSION THREAD

**Discussion by savchenko** [neovim/neovim#27042] [Q&A]
Things look solidly grim at this point. On the edge of reporting a bug against `vim.lsp.buf.format()`.

---

I'd like to format _current_ visual selection via LSP.

`documentFormattingProvider` is present in `client.server_capabilities` and formatting _the whole file_ works just fine.

What I have tried so far:
- Using `vim.lsp.buf.format()` as per documented. It fails to act on visual selection when called without parameters.
- Force `range` and `bufnr`. Same result, formats the whole buffer.
- Calling it via user command. Still formats the whole buffer.

Tested with `ruff`, `shfmt` and `autopep8` acting as formatters.

Any ideas how to make it work? As an absolutely last resort, I am considering extracting text via Lua function, passing it to an external formatter via `stdin` and then replacing selection with its `stdout`, but... This is too much, there has to be a better way.

**Comment by wookayin** [comment]
If you can share some code snippets or minimal repro you've tried, that'd be helpful. If you pass the current visual selection range into the `range` table, it should work I believe.

**Reply by savchenko** [reply]
Sure. Let's start with a function call:

```lua
vim.keymap.set('v', '\\q', function() vim.lsp.buf.format() end, { buffer = true })
```

...formats the whole file, while the documentation states otherwise:
https://github.com/neovim/neovim/blob/4615d46f93fb0b15a427f428bb1e652b4ccb873b/runtime/lua/vim/lsp/buf.lua#L195

Mapping bug? Let's try without parenthesis:

```lua
vim.keymap.set('v', '\\q', vim.lsp.buf.format, { buffer = true })
```

No, still formats the whole file. And what if we are being _really_ explicit?

```lua
local c_buf = vim.api.nvim_get_current_buf()
local s_start = vim.api.nvim_buf_get_mark(0, "<")
local s_end = vim.api.nvim_buf_get_mark(0, ">")

vim.keymap.set('v', '\\q', function() vim.lsp.buf.format({ bufnr = c_buf, range = {["start"] = s_start, ["end"] = s_end} }) end, { buffer = true })
```

...then it still formats the whole file.

**Reply by darukutsu** [reply]
Have you come up with something?

**Reply by savchenko** [reply]
Unfortunately, no. Resorted to reformatting the whole file each time.

**Reply by darukutsu** [reply]
I don't get it for example [gopls](https://github.com/golang/tools/blob/98aa9a7d0bbb11ce97c8bd9e9112db87b19ecda5/gopls/doc/design/design.md?plain=1#L354) should have `textDocument/rangeFormatting` but when calling format it does absolutely nothing on range. Even tho https://github.com/neovim/neovim/issues/27293 was implemented. Maybe typo `s/ranges/range`, what do you think? Because there are occurrences of both in neovim code-base.

**Reply by shushtain** [reply]
Do you have `rangeFormatting = { dynamicRegistration = true, rangesSupport = true }` (probably, only the second one matters) for the server that supports it in your `:checkhealth vim.lsp`? With:

```lua
vim.keymap.set("x", "<Leader>pf", vim.lsp.buf.format, { desc = "LSP : Format" })
```

I either get a formatted range, just as expected, or "no supported servers". I don't get whole buffer formatting at all, if the shortcut is called in visual mode, and the server doesn't support ranges. `emmylua_ls` and `ruff` worked for me. I'm on `v0.11.5`.
